[Cloud Security] CDR Data View versioning and migration logic

opauloh opened the pull request:
## Summary

It fixes https://github.com/elastic/kibana/issues/234835

Implemented a Data View versioning and migration system for CSP (Cloud Security Posture) Misconfiguration and Vulnerabilities to handle upgrades from legacy and old dataviews to the latest, solving issues where old environments kept stale Data Views with incorrect index patterns.

### Original issue

- Old environments referenced deprecated Data Views with stale index patterns
- CSP initialization didn't delete old or legacy Data View IDs, causing existing spaces to keep stale Data Views indefinitely
- Users upgrading from older versions experienced 'no data' issues and limited field availability

## Implementation Details

### Data View Versioning (v2)
- Added v2 suffix to both misconfigurations and vulnerabilities Data Views
- Maintained backward compatibility with v1 constants for existing code

### Array-based Migration System
- Refactored to use arrays (`CDR_*_DATA_VIEW_ID_PREFIX_OLD_VERSIONS`) for managing multiple old versions
- Added arrays (`CDR_*_DATA_VIEW_ID_PREFIX_LEGACY_VERSIONS`) for managing multiple legacy global dataviews
- Supports users skipping intermediate versions during upgrades (e.g., v1→v3)
- Automatically iterates through and removes all old Data View versions


### DataView Migration Logic

The data view migration is split into two parts:

1. Deletion of old and legacy data views during the plugin initialization (only runs once when the CSP package is installed or when Kibana is started)
2. Creation of new data views when the user navigates to the CSP page (the check runs every time the user navigates to the CSP page to see if the data views need to be created)


### Tests
   - Test deletion from v1 to current version (with space suffix)
   - Test deletion from legacy to current version (global to space-specific)
   - Test deletion of old and legacy data views during plugin initialization
   - Test creation of new data views when the user navigates to the CSP page

## Screenshot

<img width="2334" height="1522" alt="image" src="https://github.com/user-attachments/assets/a6ec652a-ef2f-4f90-8a82-9ce83b9ce831" />

<img width="1971" height="1529" alt="image" src="https://github.com/user-attachments/assets/797b58b0-5fbe-4cac-8cbd-a2c44e82b2a3" />


<img width="2269" height="1564" alt="image" src="https://github.com/user-attachments/assets/ff9beded-1bcd-4faa-b0c9-5194764d2991" />

### Example legacy dataviews (< 8.17)

<img width="1664" height="177" alt="image" src="https://github.com/user-attachments/assets/6f66888a-c70f-4fdb-af20-5c5bd9b688dd" />



## How to test this PR

To test this PR from a blank instance:
- Create a old or legacy dataview (ex. dataview with id `security_solution_cdr_latest_misconfigurations-default`) 
- Install CSPM or KSPM or CNVM integration (this will trigger Dataview migration)
- Access the Findings page (This will trigger new Dataview creation)
- Access the DataViews page and confirm the old dataview is gone, and the new versioned ids is present
- Access the findings pages and Cloud Security Posture and Vulnerabilities dashboards and confirm everything works properly

<!--ONMERGE {"backportTargets":["8.18","8.19","9.1","9.2"]} ONMERGE-->

opauloh committed:
[Cloud Security] Implement Data View versioning and migration system

## Summary
Implemented a robust Data View versioning and migration system for CSP (Cloud Security Posture) to handle upgrades from any old version to the latest, solving issues where old environments kept stale Data Views with incorrect index patterns.

## Problem Solved
- Old environments referenced deprecated Data Views with stale index patterns
- CSP initialization didn't migrate Data View IDs, causing existing spaces to keep stale Data Views indefinitely
- Users upgrading from older versions experienced 'no data' issues and limited field availability

## Implementation Details

### 1. Data View Versioning (v2)
- Added v2 suffix to both misconfigurations and vulnerabilities Data Views
- Maintained backward compatibility with v1 constants for existing code

### 2. Array-based Migration System
- Refactored to use arrays (CDR_*_DATA_VIEW_ID_PREFIX_OLD_VERSIONS) for managing multiple old versions
- Supports users skipping intermediate versions during upgrades (e.g., v1→v3)
- Automatically iterates through and removes all old Data View versions

### 3. Migration Logic
- Created migrateOldDataViews function that handles multiple versions
- Runs on every CSP route access via setupCdrDataViews
- Space-aware migration across all Kibana spaces
- Idempotent - safe to run multiple times

### 4. Comprehensive Testing
- Added tests for single version migration
- Added tests for multiple version migration scenarios
- Tests for fresh installations (no migration needed)
- Multi-space migration tests

### Files Modified
- x-pack/solutions/security/packages/kbn-cloud-security-posture/common/constants.ts
- x-pack/solutions/security/plugins/cloud_security_posture/server/saved_objects/data_views.ts
- x-pack/solutions/security/plugins/cloud_security_posture/common/constants.ts
- x-pack/solutions/security/plugins/cloud_security_posture/public/pages/vulnerabilities/vulnerabilities.tsx
- x-pack/solutions/security/test/cloud_security_posture_functional/data_views/data_views.ts
- x-pack/solutions/security/plugins/cloud_security_posture/README.md

## Development Environment
- IDE: Cursor IDE
- AI Model: Claude 3.5 Sonnet (via Cursor)
- Development approach: AI-assisted pair programming with iterative refinement

## Future Maintenance
When creating new Data View versions:
1. Add old version to the OLD_VERSIONS array
2. Update main constant to new version (e.g., _v3)
3. Migration automatically handles all upgrade paths

Fixes: elastic/security-team#10683

opauloh committed:
Removes deprecated data view ID constants

Removes the V1 data view ID constants and directly uses the first element of the OLD_VERSIONS array in the tests.
This simplifies the code and aligns it with the intended usage of managing data view migrations.

opauloh committed:
[Cloud Security] Update README with array-based data view migration instructions

- Updated documentation to reflect the new array-based migration system
- Added clear examples showing how to add versions to the OLD_VERSIONS array
- Documented the automatic migration features (skip-version support, space-aware, idempotent)
- Provided a complete example for moving from v2 to v3
- Corrected the file path to packages/kbn-cloud-security-posture/common/constants.ts
- Emphasized that no code changes are needed in the migration logic when updating versions

opauloh committed:
Updates cloud security posture data view

Updates the cloud security posture data view to version 3.

This change increments the version of the CDR misconfigurations
data view ID prefix to 'security_solution_cdr_latest_misconfigurations_v3'.
The migration logic automatically handles all versions in the array.

opauloh committed:
Removes unnecessary comments

Removes comments that are redundant and do not add value to the code.

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint_all_files --no-cache --fix'

opauloh committed:
Adds space to data view creation and navigation

Ensures data views are created within the correct space
during testing and navigation to findings and
vulnerabilities pages includes the space parameter,
triggering migration within the correct context.

opauloh committed:
Improves cloud security posture data view tests

Refactors cloud security posture data view tests to improve reliability and reduce flakiness.
This includes:
- Wrapping the forceLogout call in a try-catch block to prevent test failures due to intermittent logout issues.
- Navigating directly to the findings/dashboard page in the test space to avoid unnecessary steps and potential issues with space switching.
- Retrying forceLogout before login with read user, to be sure user is logged out.
- Using kibanaServer.request API to create old v1 data views.

opauloh committed:
Merge branch 'main' into csp/migrate-dataviews

animehart reviewed (approved):
LGTM
Tested locally

opauloh committed:
Migrates CDR data views across all spaces

Migrates existing Cloud Discovery and Response (CDR) data views across all spaces to ensure consistent data presentation after upgrades.

This change removes the old data view migration logic which was space-dependent and replaces it with a process that iterates through all spaces, identifies old data views based on predefined prefixes, and migrates them.

It also addresses a previous limitation where data view migration was only performed within the current space, potentially leaving outdated data views in other spaces. This ensures a seamless upgrade experience for users with multi-space configurations.

opauloh committed:
Removes Spaces service dependency for CSP plugin

Refactors the Cloud Security Posture (CSP) plugin to remove its
dependency on the Spaces service when migrating data views.

This change simplifies the initialization process and data view
migration logic by removing the need to fetch all spaces, instead
searching for data views across all namespaces. This resolves
an issue where the migration could fail if the Spaces service was
unavailable or if the number of spaces exceeded the page limit.

opauloh committed:
Adds CSP data view saved object tests

Adds unit tests for CSP data view saved object functions:
- installDataView: Tests data view creation and error handling.
- migrateCdrDataViewsForAllSpaces: Tests migration and deletion of old data views across spaces.
- setupCdrDataViews: Tests the installation of both misconfigurations and vulnerabilities data views.

IDE: Cursor
AI Model: Claude Sonnet 4.5

opauloh committed:
Migrates CSP data views and cleans up legacy versions

Adds logic to migrate Cloud Security Posture data views from older and legacy versions to the current version.

This change introduces legacy data view IDs for migration and updates the migration logic to handle both legacy (global) and versioned (space-specific) data views.  It ensures that old data views are properly migrated and removed during plugin initialization, supporting upgrades from older versions and fresh installations. The migration system is also updated to be space-aware and idempotent, ensuring safe and consistent migration across all Kibana spaces.

Adds tests to cover migration from v1 and legacy versions and across multiple spaces.

opauloh committed:
Merge branch 'main' into csp/migrate-dataviews

opauloh committed:
Ensures data view migration does not delete others

Adds a test case to verify that data view migration does not
accidentally delete unrelated data views. This ensures that the
migration process only targets the intended data views for deletion,
preventing any unintended data loss.

opauloh committed:
Merge remote-tracking branch 'origin/csp/migrate-dataviews' into csp/migrate-dataviews

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/capture_oas_snapshot   --include-path /api/status   --include-path /api/alerting/rule/   --include-path /api/alerting/rules   --include-path /api/actions   --include-path /api/security/role   --include-path /api/spaces   --include-path /api/streams   --include-path /api/fleet   --include-path /api/saved_objects/_import   --include-path /api/saved_objects/_export   --include-path /api/maintenance_window   --include-path /api/agent_builder --update'

opauloh committed:
Merge branch 'main' into csp/migrate-dataviews

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/351980)
* Commit: 471fab5a04eb2fce73708203d3dde5f16dc503e2
* [Cloud Deployment](https://kibana-pr-238547-17bb0f.kb.us-west2.gcp.elastic-cloud.com)

### History
* :green_heart: [Build #351963](https://buildkite.com/elastic/kibana-pull-request/builds/351963) succeeded 58c5f953a584742ea9fe03c94a02f90697d5efa6
* :broken_heart: [Build #351958](https://buildkite.com/elastic/kibana-pull-request/builds/351958) failed 827ebbf0e43d9d03d42accde130a045dc3438135
* :yellow_heart: [Build #350959](https://buildkite.com/elastic/kibana-pull-request/builds/350959) was flaky 41584a991b8d87bcdd38ed01e706945397fd36ad
* :green_heart: [Build #350014](https://buildkite.com/elastic/kibana-pull-request/builds/350014) succeeded ad2ac6a5c0b2beaed09ba72f9f8d1c847e3069f2
* :broken_heart: [Build #349947](https://buildkite.com/elastic/kibana-pull-request/builds/349947) failed 3c98a0acb3eb031be184d82f213bb4182d7839c0
* :broken_heart: [Build #349938](https://buildkite.com/elastic/kibana-pull-request/builds/349938) failed ebdbd4f6491b6a399cab809727fb335ca0e182a3

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/351980","number":351980,"commit":"471fab5a04eb2fce73708203d3dde5f16dc503e2"},{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/351963","number":351963,"commit":"58c5f953a584742ea9fe03c94a02f90697d5efa6"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/351958","number":351958,"commit":"827ebbf0e43d9d03d42accde130a045dc3438135"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/350959","number":350959,"commit":"41584a991b8d87bcdd38ed01e706945397fd36ad"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/350014","number":350014,"commit":"ad2ac6a5c0b2beaed09ba72f9f8d1c847e3069f2"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/349947","number":349947,"commit":"3c98a0acb3eb031be184d82f213bb4182d7839c0"}],"number":351980}
buildkite-pr-comment-->

opauloh committed:
Deletes old and legacy CSP data views

Removes old and legacy Cloud Security Posture data views during plugin initialization to ensure a clean state and prevent conflicts with new data view structures.

This change ensures that only the current data views are available, simplifying data management and improving performance. The data view deletion logic runs during plugin initialization.

opauloh committed:
Merge branch 'main' into csp/migrate-dataviews

opauloh committed:
Refactors data view creation in functional tests

Introduces a `createDataView` helper function to simplify the creation of data views in functional tests.

This change improves code readability and reduces duplication, especially in CSP migration tests, by encapsulating the data view creation logic.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/352314)
* Commit: 743cc67d855506555007d36d81b791131584005d

### Failed CI Steps
* [Linting (with types)](https://buildkite.com/elastic/kibana-pull-request/builds/352314#0199f3ab-0c4a-41c3-8953-36d38c692058)

### History
* :broken_heart: [Build #352296](https://buildkite.com/elastic/kibana-pull-request/builds/352296) failed 3251a2bf1403cdbe42ffdd65cf3d70abdddad149
* :broken_heart: [Build #351980](https://buildkite.com/elastic/kibana-pull-request/builds/351980) failed 471fab5a04eb2fce73708203d3dde5f16dc503e2
* :green_heart: [Build #351963](https://buildkite.com/elastic/kibana-pull-request/builds/351963) succeeded 58c5f953a584742ea9fe03c94a02f90697d5efa6
* :broken_heart: [Build #351958](https://buildkite.com/elastic/kibana-pull-request/builds/351958) failed 827ebbf0e43d9d03d42accde130a045dc3438135
* :yellow_heart: [Build #350959](https://buildkite.com/elastic/kibana-pull-request/builds/350959) was flaky 41584a991b8d87bcdd38ed01e706945397fd36ad

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/352314","number":352314,"commit":"743cc67d855506555007d36d81b791131584005d"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/352296","number":352296,"commit":"3251a2bf1403cdbe42ffdd65cf3d70abdddad149"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/351980","number":351980,"commit":"471fab5a04eb2fce73708203d3dde5f16dc503e2"},{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/351963","number":351963,"commit":"58c5f953a584742ea9fe03c94a02f90697d5efa6"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/351958","number":351958,"commit":"827ebbf0e43d9d03d42accde130a045dc3438135"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/350959","number":350959,"commit":"41584a991b8d87bcdd38ed01e706945397fd36ad"}],"number":352314}
buildkite-pr-comment-->

opauloh committed:
fix type checking error

opauloh committed:
Updates CSP package installation flow

Updates the CSP data views functional tests to install the CSP package and package policy.
The installation now includes creating an agent policy and a package policy,
which is necessary for the data views to be created correctly.
Also updates the tests to use the plugin version number and adds space context for api calls.

opauloh committed:
Merge branch 'main' into csp/migrate-dataviews

opauloh committed:
fix type checking

kibanamachine commented:
## 💔 Some backports could not be created

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.18|Backport failed because of merge conflicts|
|❌|8.19|Backport failed because of merge conflicts|
|❌|9.1|Backport failed because of merge conflicts|
|✅|9.2|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/239974">](https://github.com/elastic/kibana/pull/239974)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 238547
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->

maxcold commented on x-pack/solutions/security/packages/kbn-cloud-security-posture/common/constants.ts:
```diff
@@ -41,8 +41,21 @@ export const STATUS_API_CURRENT_VERSION = '1';
+];
+// Array of old data view IDs for migration purposes
+// Add new deprecated versions here when updating to a new version
+export const CDR_MISCONFIGURATIONS_DATA_VIEW_ID_PREFIX_OLD_VERSIONS = [
```
@opauloh a bit confsing having legacy, old and current. What is the difference between legacy and old?

opauloh replied:
Hey Max, answered on slack, but cross posting it here in case you want to continue the discussion here or if someone stumble upon this:

The legacy dataviews are the dataviews we used to have before we introduced the CDR (The ones we used to install from saved objects from the [integration](https://github.com/elastic/integrations/tree/detection-rules/8.14.7-beta.1-bdee4d45/packages/cloud_security_posture/kibana/index_pattern) repo). I realized that when users upgrade from earlier 8.x versions to 8.19, those dataviews would remain there without purpose, so I included the deletion of them in the PR. Those IDs are static, and we won't need to update the legacy dataviews array ever again.

While the old dataviews are the older versions of the current CDR dataviews that we uses, so deprecations from now on will need only the following steps:
- Update the `CDR_MISCONFIGURATIONS_DATA_VIEW_ID_PREFIX` const with the new version.
- Update the `CDR_MISCONFIGURATIONS_DATA_VIEW_ID_PREFIX_OLD_VERSIONS` array with the older version.

The reason why I stored them separately is that the legacy ones are stored in the global namespace, so they need different handling, and also because we don't need to update that list ever again (main reason).

I also attempted to document [here](https://github.com/opauloh/kibana/blob/main/x-pack/solutions/security/plugins/cloud_security_posture/README.md#how-to-update-data-view-version) that we don't need to worry about the Legacy ones, but I realized it wasn't that clear after all :sweat_smile:. So I'll be happy to address those feedbacks in a follow up PR. If it's clearer, I can keep all legacies in a single array (not separated by misc and vuln) with a better description to not update the array

maxcold replied:
as discussed in Slack, maybe we can name it a bit less generic, smth like `pre_8_19` and add a comment expaining the need for a separate handling
